renderer/backend/GpuBuffer.cpp

```diff
@@ -0,0 +1,144 @@
+/*****************************************************************************
+                    The Dark Mod GPL Source Code
+
+ This file is part of the The Dark Mod Source Code, originally based
+ on the Doom 3 GPL Source Code as published in 2011.
+
+ The Dark Mod Source Code is free software: you can redistribute it
+ and/or modify it under the terms of the GNU General Public License as
+ published by the Free Software Foundation, either version 3 of the License,
+ or (at your option) any later version. For details, see LICENSE.TXT.
+
+ Project: The Dark Mod (http://www.thedarkmod.com/)
+
+******************************************************************************/
+#include "precompiled.h"
+#include "GpuBuffer.h"
+
+extern idCVarBool r_usePersistentMapping;
+
+const int GpuBuffer::NUM_FRAMES;
+
+void GpuBuffer::Init( GLenum type, GLuint size, GLuint alignment, byte *staticData, GLuint numStaticBytes ) {
+	assert( idMath::IsPowerOfTwo( alignment ) && "Alignment must be a power of 2" );
+	
+	if( bufferObject ) {
+		Destroy();
+	}
+
+	staticDataSize = ALIGN( numStaticBytes, alignment );
+	frameSize = ALIGN( size, alignment );
+	this->alignment = alignment;
+	this->type = type;
+
+	qglGenBuffers( 1, &bufferObject );
+	qglBindBuffer( type, bufferObject );
+
+	usesPersistentMapping = r_usePersistentMapping && GLAD_GL_ARB_buffer_storage;
+
+	totalSize = staticDataSize + NUM_FRAMES * frameSize;
+	if ( usesPersistentMapping ) {
+		qglBufferStorage( type, totalSize, nullptr, GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
+		bufferContents = ( byte* )qglMapBufferRange( type, 0, totalSize, GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
+	} else {
+		qglBufferData( type, totalSize, nullptr, GL_DYNAMIC_DRAW );
+		bufferContents = ( byte* )Mem_Alloc16( totalSize );
+	}
+
+	memcpy( bufferContents, staticData, numStaticBytes );
+	if ( !usesPersistentMapping ) {
+		qglBufferSubData( type, 0, numStaticBytes, staticData );
+	}
+	
+	currentFrame = 0;
+	bytesCommittedInCurrentFrame = 0;
+}
+
+void GpuBuffer::Destroy() {
+	if ( bufferObject == 0 ) {
+		return;
+	}
+
+	for (int i = 0; i < NUM_FRAMES; ++i) {
+		if ( frameFences[i] != nullptr ) {
+			qglDeleteSync( frameFences[i] );
+			frameFences[i] = nullptr;
+		}
+	}
+
+	if ( usesPersistentMapping ) {
+		qglBindBuffer( type, bufferObject );
+		qglUnmapBuffer( type );
+	} else {
+		Mem_Free16( bufferContents );
+	}
+
+	qglDeleteBuffers( 1, &bufferObject );
+	bufferObject = 0;
+	bufferContents = nullptr;
+}
+
+byte * GpuBuffer::CurrentWriteLocation() const {
+	return bufferContents + CurrentOffset();
+}
+
+GLuint GpuBuffer::BytesRemaining() const {
+	return frameSize - bytesCommittedInCurrentFrame;
+}
+
+void GpuBuffer::Commit( GLuint numBytes ) {
```

Comment:
Hm, actually, since the `VertexCache` is filled in the frontend which must not call GL commands at all, we'd need a little extra abstraction there, anyway. So we'd end up with a single `glBufferSubData` per frame out of necessity, and this is what's happening right now, too.

I think that handling could be done entirely in the `VertexCache`; at worst it might need an extra helper function in the `GpuBuffer`.

Reply:
Yes, I think something like inline `GpuBuffer::Align` method would be enough.

Reply:
Either way, I believe it should not require significant refactoring of the `GpuBuffer`, so I'd like to postpone this to a later PR when we might actually refactor `VertexCache` to use this class :)